README.md
name_service_lookups
====================

[![Build
Status](https://travis-ci.org/wcooley/puppet-name_service_lookups.png?branch=master)](https://travis-ci.org/wcooley/puppet-name_service_lookups)

This module is a collection of Puppet functions for system database (NSS) lookups,
of the `get*by*` type.

Included databases:
 * passwd
 * group
 * hosts
 * services

getpwnam
--------
Looks up a user by name. See `getpwnam(3)`.

Given a username as parameter, it returns a hash with the keys:
  * name
  * passwd
  * uid
  * gid
  * gecos
  * dir
  * shell

  Returns *undef* is no such username exists.

*Example:*

      getpwnam('root')

Would return (approximately):


      {
        'name':     'root',
        'passwd':   '*',
        'uid':      0,
        'gid':      0,
        'gecos':    'root',
        'dir':      '/root',
        'shell':    '/bin/bash'
      }

- *Type*: rvalue

getpwuid
--------
Looks up a user by UID. See `getpwuid(3)`.

Given a user ID as parameter, it returns a hash with the keys:
  * name
  * passwd
  * uid
  * gid
  * gecos
  * dir
  * shell

  Returns *undef* is no such user ID exists.

*Example:*

    getpwuid(0)

Would return (approximately):

    {
        'name':     'root',
        'passwd':   '*',
        'uid':      0,
        'gid':      0,
        'gecos':    'root',
        'dir':      '/root',
        'shell':    '/bin/bash'
    }

- *Type*: rvalue

getgrgid
--------
Looks up a group by GID. See `getgrgid(3)`.

Given a GID as parameter, it returns a hash with keys:
  * name - The name of the group.
  * passwd - A hashed group password (usually only '*').
  * gid - Group ID (GID).
  * mem - An array of the names of users who are members of the group.

Returns *undef* if no such group ID exists.

*Example:*

        getgrgid(0)

Would result in (approximately):

        {
            'name': 'root',
            'passwd': '*',
            'gid': 0,
            'mem': ['root']
        }

- *Type*: rvalue

getgrnam
--------
Looks up a group by name. See `getgrnam(3)`.

Given a group name as parameter, it returns a hash with keys:
 * name - The name of the group.
 * passwd - A hashed group password (usually only '*').
 * gid - Group ID (GID).
 * mem - An array of the names of users who are members of the group.

Returns *undef* if no such group exists.

*Example:*

    getgrnam('root')

Would return (approximately):

    {
        'name': 'root',
        'passwd': '*',
        'gid': 0,
        'mem': ['root']
    }

- *Type*: rvalue

gethostbyaddr
-------------
Looks up a host by address. See `gethostbyaddr(3)`.

Given an IP address as parameter, it returns a hash with keys:
  * name - The host name.
  * aliases - An array of aliases for the host name.
  * address - The address of the host.

*Example:*

    gethostbyaddr('127.0.0.1')

Would return (approximately):

    {
        'name': 'localhost'
        'aliases': ["1.0.0.127.in-addr.arpa"]
        'address': '127.0.0.1'
    }

- *Type*: rvalue

gethostbyname
-------------
Looks up a host by address. See `gethostbyname(3)`.

Given a host name as a parameter, it returns a hash with keys:
 * name - The host name.
 * aliases - An array of aliases for the host name.
 * address - The address of the host.

    Returns *undef* if no such host exists.

*Example:*

    gethostbyname('localhost')

Would return (approximately):

    {
        'name': 'localhost',
        'aliases': ['localhost.localdomain', '1.0.0.127.in-addr.arpa'],
        'address': '127.0.0.1'
    }

- *Type*: rvalue

getservbyname
-------------
Looks up a service by name. See `getservbyname(3)`.

Given a network service name (such as from `/etc/services`) and optionally a
transport protocol, it returns the associated port number. If a protocol is not
specified, the default 'tcp' is used.

Returns *undef* is no such service name exists.

*Example:*

    getservbyname('syslog', 'udp')

Would return (approximately):

    514

*Example:*

    getservbyname('syslog')

Might return:

    undef

- *Type*: rvalue

getservbyport
-------------
Looks up a service by port number. See `getservbyport(3)`. Note that this is
only available on Ruby 1.9 and later.

Given a network service port (such as from `/etc/services`) and optionally a
transport protocol, it returns the associated service name. If a protocol is
not specified, the default 'tcp' is used.

Returns *undef* is no such port number exists.

*Example:*

    getservbyport(514)

Would return (approximately):

    shell

*Example:*

    getservbyname(514, 'udp')

Would return:

    syslog

- *Type*: rvalue


Author
-------

Wil Cooley wcooley(at)nakedape.cc

Support
-------

Please log tickets and issues at our [Projects
site](https://github.com/wcooley/puppet-name_service_lookups).

License
-------
    Author:: Wil Cooley (wcooley(at)nakedape.cc)
    Copyright:: Copyright (c) 2013 Wil Cooley
    License:: Apache License, Version 2.0

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG.md
2014-08-23 - Release 0.1.0
--------------------------

Features:
 * Initial release. Includes support for the following functions:
    * getpwnam
    * getpwuid
    * getgrnam
    * getgrgid
    * gethostbyaddr
    * gethostbyname
    * getservbyname
    * getservbyport - (Ruby 1.9+ only)


spec/spec_helper.rb
# encoding: UTF-8

require 'puppetlabs_spec_helper/module_spec_helper'


spec/functions/getgrgid_spec.rb
# encoding: UTF-8

require 'spec_helper'

describe 'getgrgid' do

  it 'should exist' do
    Puppet::Parser::Functions.function('getgrgid') \
      .should eq 'function_getgrgid'
  end
  it 'should return expected result with valid input' do
    grent_ary = [
      ['name', 'jensenb'],
      ['passwd', '*'],
      ['gid', 101],
      ['mem', ['wcooley']],
    ]
    grent_vals = grent_ary.collect { |k,v| v }
    grent_hash = Hash[grent_ary]

    grent_struct = Struct::Group[*grent_vals]

    Etc.expects(:getgrgid).with(101).returns(grent_struct)
    expect(subject).to run.with_params(101).and_return(grent_hash)
  end

  it 'should return undef on stubbed lookup failure' do
    Etc.expects(:getgrgid).with(1010) \
      .raises(ArgumentError, "can't find group for 1010")
    expect(subject).to run.with_params(1010).and_return(:undef)
  end

  it 'should re-raise exceptions other than lookup failure' do
    Etc.expects(:getgrgid).with(0101) \
      .raises(ArgumentError, 'not the right argument')
    expect(subject).to run.with_params(0101) \
      .and_raise_error(ArgumentError, 'not the right argument')
  end

  it 'should raise error for non-integer input' do
    expect(subject).to run.with_params('not-an-integer') \
      .and_raise_error(ArgumentError, /invalid value for Integer/)
  end
end


spec/functions/getgrnam_spec.rb
# encoding: UTF-8

require 'spec_helper'

describe 'getgrnam' do
  it 'should exist' do
    Puppet::Parser::Functions.function('getgrnam') \
      .should eq 'function_getgrnam'
  end

  it 'should return expected result with valid input' do
    grent_ary = [
      ['name', 'jensenb'],
      ['passwd', '*'],
      ['gid', 101],
      ['mem', ['wcooley']],
    ]
    grent_vals = grent_ary.collect { |k,v| v }
    grent_hash = Hash[grent_ary]

    grent_struct = Struct::Group[*grent_vals]

    Etc.expects(:getgrnam).with('jensenb').returns(grent_struct)

    #scope = PuppetlabsSpec::PuppetInternals.scope
    #expect(scope.function_getgrnam(['jensenb'])).to eq(grent_hash)
    expect(subject).to run.with_params('jensenb').and_return(grent_hash)
  end

  it 'should return undef on stubbed lookup failure' do
    Etc.expects(:getgrnam).with('zoot') \
      .raises(ArgumentError, "can't find group for zoot")
    expect(subject).to run.with_params('zoot').and_return(:undef)
  end

  it 'should re-raise exceptions other than lookup failure' do
    Etc.expects(:getgrnam).with('root') \
      .raises(ArgumentError, 'not the right argument')
    expect(subject).to run.with_params('root') \
      .and_raise_error(ArgumentError, 'not the right argument')
  end

end


lib/puppet/parser/functions/getgrgid.rb
# encoding: UTF-8

require 'etc'

module Puppet::Parser::Functions
  newfunction(:getgrgid, :type => :rvalue, :doc => <<-EOS) do |args|
Looks up a group by GID. See `getgrgid(3)`.

Given a GID as parameter, it returns a hash with keys:
  * name - The name of the group.
  * passwd - A hashed group password (usually only '*').
  * gid - Group ID (GID).
  * mem - An array of the names of users who are members of the group.

Returns *undef* if no such group ID exists.

*Example:*

    getgrgid(0)

Would result in (approximately):

    {
      'name': 'root',
      'passwd': '*',
      'gid': 0,
      'mem': ['root']
    }

    EOS

    gid = Integer(args[0])
    begin
      grent = Etc.getgrgid(gid)
      grvals = grent.values[0..3]
      grkeys = grent.members[0..3].collect { |k| k.to_s }
      result = Hash[grkeys.zip(grvals)]

    rescue ArgumentError => ae
      if ae.to_s =~ /can't find group/
        result = :undef
      else
        raise ae
      end
    end
  end
end



lib/puppet/parser/functions/getgrnam.rb
# encoding: UTF-8

require 'etc'

module Puppet::Parser::Functions
  newfunction(:getgrnam, :type => :rvalue, :doc => <<-EOS) do |args|
Looks up a group by name. See `getgrnam(3)`.

Given a group name as parameter, it returns a hash with keys:
  * name - The name of the group.
  * passwd - A hashed group password (usually only '*').
  * gid - Group ID (GID).
  * mem - An array of the names of users who are members of the group.

Returns *undef* if no such group exists.

*Example:*

    getgrnam('root')

Would return (approximately):

    {
      'name': 'root',
      'passwd': '*',
      'gid': 0,
      'mem': ['root']
    }

    EOS

    grp = args[0]

    begin
      grent = Etc.getgrnam(grp)
      grvals = grent.values[0..3]

      grkeys = grent.members[0..3].collect { |k| k.to_s }

      result = Hash[grkeys.zip(grvals)]
    rescue ArgumentError => ae
      if ae.to_s =~ /can't find group/
        result = :undef
      else
        raise ae
      end
    end

    result
  end
end


lib/puppet/parser/functions/getpwnam.rb
# encoding: UTF-8

require 'etc'

module Puppet::Parser::Functions
  newfunction(:getpwnam, :type => :rvalue, :doc => <<-EOS) do |args|
Looks up a user by name. See `getpwnam(3)`.

Given a username as parameter, it returns a hash with the keys:
  * name
  * passwd
  * uid
  * gid
  * gecos
  * dir
  * shell

  Returns *undef* is no such username exists.

*Example:*

  getpwnam('root')

Would return (approximately):
  {
    'name':     'root',
    'passwd':   '*',
    'uid':      0,
    'gid':      0,
    'gecos':    'root',
    'dir':      '/root',
    'shell':    '/bin/bash'
  }
EOS
    user = args[0]

    begin
      pwent = Etc.getpwnam(user)

      # Truncate to first 7 for consistency between platforms
      pwvals = pwent.values[0..6]

      # In Ruby 1.8, members returns strings but in 1.9 returns symbols
      pwkeys = pwent.members[0..6].collect { |k| k.to_s }

      result = Hash[pwkeys.zip(pwvals)]

    rescue ArgumentError => ae
      if ae.to_s =~ /can't find user/
       result = :undef
      else
        raise ae
      end
    end

    result
  end
end


lib/puppet/parser/functions/gethostbyaddr.rb
# encoding: UTF-8

require 'ipaddr'
require 'socket'

module Puppet::Parser::Functions
  newfunction(:gethostbyaddr, :type => :rvalue, :doc => <<-EOS) do |args|
Looks up a host by address. See `gethostbyaddr(3)`.

Given an IP address as parameter, it returns a hash with keys:
  * name - The host name.
  * aliases - An array of aliases for the host name.
  * address - The address of the host.

*Example:*

    gethostbyaddr('127.0.0.1')

Would return (approximately):

    {
      'name': 'localhost'
      'aliases': ["1.0.0.127.in-addr.arpa"]
      'address': '127.0.0.1'
    }

    EOS

    address = IPAddr.new(args[0])

    begin
      hostent = Socket.gethostbyaddr(address.hton)
      result = { 'name' => hostent[0], 'aliases' => hostent[1] }
      result['address'] = IPAddr.new_ntoh(hostent[3]).to_s

    rescue SocketError => se
      if se.to_s == 'host not found'
        result = :undef
      else
        raise se
      end
    end

    result
  end
end


lib/puppet/parser/functions/gethostbyname.rb
# encoding: UTF-8

require 'ipaddr'
require 'socket'

module Puppet::Parser::Functions
  newfunction(:gethostbyname, :type => :rvalue, :doc => <<-EOS) do |args|
Looks up a host by address. See `gethostbyname(3)`.

Given a host name as a parameter, it returns a hash with keys:
    * name - The host name.
    * aliases - An array of aliases for the host name.
    * address - The address of the host.

    Returns *undef* if no such host exists.

*Example:*

    gethostbyname('localhost')

Would return (approximately):

  {
    'name': 'localhost',
    'aliases': ['localhost.localdomain', '1.0.0.127.in-addr.arpa'],
    'address': '127.0.0.1'
  }

    EOS

    name = args[0]

    begin
      hostent = Socket.gethostbyname(name)
      result = { 'name' => hostent[0], 'aliases' => hostent[1] }
      result['address'] = IPAddr.new_ntoh(hostent[3]).to_s

    rescue SocketError => se
      if se.to_s =~ %r{not known}
        result = :undef
      else
        raise se
      end
    end

    result
  end
end


lib/puppet/parser/functions/getservbyname.rb
# encoding: UTF-8

require 'socket'

module Puppet::Parser::Functions
  newfunction(:getservbyname, :type => :rvalue, :doc => <<-EOS) do |args|
Looks up a service by name. See `getservbyname(3)`.

Given a network service name (such as from `/etc/services`) and optionally a
transport protocol, it returns the associated port number. If a protocol is not
specified, the default 'tcp' is used.

Returns *undef* is no such service name exists.

*Example:*

  getservbyname('syslog', 'udp')

Would return (approximately):

  514

*Example:*

  getservbyname('syslog')

Might return:

  undef


EOS
    begin
      result = Socket.getservbyname(*args)

    rescue SocketError => se
      if se.to_s =~ %r{no such service}
        result = :undef
      else
        raise se
      end
    end

    result
  end
end


lib/puppet/parser/functions/getservbyport.rb
# encoding: UTF-8

require 'socket'

module Puppet::Parser::Functions
  newfunction(:getservbyport, :type => :rvalue, :doc => <<-EOS) do |args|
Looks up a service by port number. See `getservbyport(3)`. Note that this is
only available on Ruby 1.9 and later.

Given a network service port (such as from `/etc/services`) and optionally a
transport protocol, it returns the associated service name. If a protocol is
not specified, the default 'tcp' is used.

Returns *undef* is no such port number exists.

*Example:*

  getservbyport(514)

Would return (approximately):

  shell

*Example:*

  getservbyname(514, 'udp')

Would return:

  syslog


EOS

    args[0] = Integer(args[0])

    begin
      result = Socket.getservbyport(*args)
    rescue SocketError => se
      if se.to_s =~ %r{no such service}
        result = :undef
      else
        raise se
      end
    end

    result
  end
end



lib/puppet/parser/functions/getpwuid.rb
# encoding: UTF-8

require 'etc'

module Puppet::Parser::Functions
  newfunction(:getpwuid, :type => :rvalue, :doc => <<-EOS) do |args|
Looks up a user by UID. See `getpwuid(3)`.

Given a user ID as parameter, it returns a hash with the keys:
  * name
  * passwd
  * uid
  * gid
  * gecos
  * dir
  * shell

  Returns *undef* is no such user ID exists.

*Example:*

  getpwuid(0)

Would return (approximately):
  {
    'name':     'root',
    'passwd':   '*',
    'uid':      0,
    'gid':      0,
    'gecos':    'root',
    'dir':      '/root',
    'shell':    '/bin/bash'
  }
EOS

    user = Integer(args[0])

    begin
      pwent = Etc.getpwuid(user)

      # Truncate to first 7 for consistency between platforms
      pwvals = pwent.values[0..6]

      # In Ruby 1.8, members returns strings but in 1.9 returns symbols
      pwkeys = pwent.members[0..6].collect { |k| k.to_s }

      result = Hash[pwkeys.zip(pwvals)]

    rescue ArgumentError => ae
      if ae.to_s =~ /can't find user/
       result = :undef
      else
        raise ae
      end
    end

    result
  end
end
